src/osd/ReplicatedBackend.cc

```diff
@@ -1269,6 +1269,21 @@ void ReplicatedBackend::calc_head_subsets(
   uint64_t size = obc->obs.oi.size;
   if (size)
     data_subset.insert(0, size);
+  else
+    return;
+  if (parent->min_peer_features() & CEPH_OSD_PARTIAL_RECOVERY) {
+    map<hobject_t, pg_missing_t::item, hobject_t::ComparatorWithDefault>::const_iterator it
+      = missing.missing.find(head);
+    assert(it != missing.missing.end());
+    interval_set<uint64_t> unmodified;
+    unmodified.insert(0, size);
+    unmodified.intersection_of(it->second.dirty_extents);
+    data_subset.subtract(unmodified);
```

Comment:
Based on current strategy, the interval (new_size ~ old_size) of truncate operations might not add to dirty_extents since we always consider the data (from new_size to old_size) in the old object is invalid based on its data_subset. Therefore, the invalid data will not be cloned to the recovery object as expected.